### ISSUE TITLE
Question: hooking serialization of specific types

### ISSUE THREAD

**Issue by kentcb** [google/built_value.dart#634]
Hi,

I couldn't find anything on this in the docs. Sorry if I've missed it.

What is the correct mechanism to use if I want to supplement the serialization of a specific type? For example, when serializing a type T, I want to translate one of its fields into a different format and no longer store the original value. Then, during deserialization, I want to perform the reverse translation. Ideally, I could do this without making the field `@nullable` or having to mark it with `@BuiltValueField(serialize: false)` (because the custom serialization logic will take precedence and gets the final say on whether it's serialized anyway).

I saw the various `Serializer` interfaces, but it's not clear to me which I should use, and how. I was hoping to have a hooks where I get a `Map<String, dynamic>`, which I can then massage as I see fit.

How can I achieve this?

**Comment by kentcb** [comment]
Hmm, I really find it confusing that `StructuredSerializer<T>` gives you an `Iterable` instead of a `Map`. Seems to make this difficult to work with, or perhaps I'm just going about this all wrong...

**Comment by davidmorgan** [comment]
There are two mechanisms: `SerializerPlugin` runs before/after all serializations/deserializations; or, you can replace the generated serializer for a particular type with your own serializer, which can do anything it likes.

`StandardJsonSerializerPlugin` is an example of a plugin; it converts built_value's default list-based format into a map-based one.

**Comment by kentcb** [comment]
Hmm, OK. The docs don't explain how `SerializerPlugin` works (still experimental) but I tried using iterables of raw data with it and it seems to want objects. So it doesn't align with my use case.

Here's an example of what I ended up with by using `StructuredSerializer<T>`:

```dart
// serialize:   itemSourcePath -> sourcePath
// deserialize: sourcePath -> itemSourcePath
class ProductItemSourceDtoSerializer implements StructuredSerializer<ProductItemSourceDto> {
  @override
  Iterable<Type> get types => [ProductItemSourceDto];

  @override
  String get wireName => "ProductItemSourceDto";

  StructuredSerializer<ProductItemSourceDto> get defaultSerializer => ProductItemSourceDto.serializer as StructuredSerializer<ProductItemSourceDto>;

  @override
  ProductItemSourceDto deserialize(Serializers serializers, Iterable serialized, {FullType specifiedType = FullType.unspecified}) {
    final serializedList = serialized.toList();

    // sourcePath (DocumentReference) -> itemSourcePath (String)
    for (var i = 0; i < serializedList.length; i += 2) {
      if (serializedList[i] == "sourcePath") {
        final DocumentReference sourcePath = serializedList[i + 1];
        serializedList[i] = "itemSourcePath";
        serializedList[i + 1] = sourcePath.path;
        break;
      }
    }

    final result = defaultSerializer.deserialize(serializers, serializedList);
    return result;
  }

  @override
  Iterable serialize(Serializers serializers, ProductItemSourceDto object, {FullType specifiedType = FullType.unspecified}) {
    final serialized = defaultSerializer.serialize(serializers, object, specifiedType: specifiedType);
    final serializedList = serialized.toList();

    // itemSourcePath (String) -> sourcePath (DocumentReference)
    for (var i = 0; i < serializedList.length; i += 2) {
      if (serializedList[i] == "itemSourcePath") {
        final String itemSourcePath = serializedList[i + 1];
        serializedList[i] = "sourcePath";
        serializedList[i + 1] = Firestore.instance.document(itemSourcePath);
        break;
      }
    }

    return serializedList;
  }
}
```

It seems a little cumbersome to me, and the `defaultSerializer` property results in a compiler warning: `Avoid using as`. Not entirely sure how to get around that other than being more wordy:

```dart
  StructuredSerializer<ProductItemSourceDto> get defaultSerializer {
    final StructuredSerializer<ProductItemSourceDto> result = ProductItemSourceDto.serializer;
    return result;
  }
```

Does this look like what you'd expect, @davidmorgan? Anything you would improve?

**Comment by davidmorgan** [comment]
"Avoid using as" is an optional lint, I would suggest turning it off as it's nonstandard :) see package:pedantic for the standard set of lints.

That looks like a reasonable solution. I agree it's not beautiful--making serialization nicely extensive is still an open question, really :)

**Comment by kentcb** [comment]
Thanks for tip on pedantic - I've just switched to that.

However, I still had a problem with code like this:

```dart
final Map<String, dynamic> result = serializers.serialize(
  foo,
  specifiedType: const FullType(FooDto),
);
```

I found that this was due to having `implicit-casts` disabled in my `analysis_options.yaml`. I had it enabled due to "advice" I found [here](https://www.dartlang.org/guides/language/analysis-options#enabling-additional-type-checks).

But with it enabled, I don't see a good way to fix the above code. If I use `as` then any type mismatch will result in a `null` rather than an exception (which I'd want in this case). If this were C#, I'd just cast:

```C#
final result = (Map<String, dynamic>)serializers.serialize(
  foo,
  specifiedType: const FullType(FooDto),
);
```

But as far as I can tell, Dart has no cast operator. Any thoughts on this? Should I just disable `implicit-casts` and be done with it?
